### PR TITLE
Lexer/Preprocessor Errors

### PR DESCRIPTION
**Pull request by est31**
As suggested in https://github.com/calixteman/rust-cpp-parser/pull/5#issuecomment-622993852

Originally created for parsing errors (allows you to do `error!` instead of `unreachable!`)

### CODE REVIEW

**Comment by calixteman** [comment]
What do you think of using failure: https://rust-lang-nursery.github.io/failure/ ?

**Comment by est31** [comment]
1. The trend seems to rather use anyhow instead of failure

2. I don't think that error handling libraries are well suited for compilers because attaching location info to the errors is non trivial and most errors come from inside the compiler anyway. Tell me if I'm wrong! Maybe we can include an anyhow cause to the `Error` struct though e.g. for IO errors.

**Comment by calixteman** [comment]
Ok.
So just add an enum, I think that for a lib it's a way better for the user to be able to filter on errors.
We could have some `lexer::Error`, `preprocessor::Error`, ... and a global Error to wrap the errors coming from lexer, pp, ...

**Comment by est31** [comment]
What different data would those errors contain? I'd just have one for both for the moment and if the need arises, split them.

**Comment by calixteman** [comment]
I think file & span is enough for now.
I think we should have errors with parameter for example "Invalid char `{:x}' at line {}..." 
I don't know if it's possible but having a format to colorize some parts of the error msg when dump on stderr, e.g "Invalid char {:x, red} at...", and having colors should be configurable.
 
Probably at some point it'd be interesting to improve location thing to be able to handle file inclusion and to be able to have a kind of stack of the different locations.

**Comment by est31** [comment]
Put stray `#endif` into ascii.cpp, do cargo run and enjoy :).

**Comment by est31** [comment]
@calixteman is this PR good? If merged, I can change #5 to use the new system and after that do even more errors.

**Comment by calixteman** [comment]
Sorry but I didn't see any answer on:
"So just add an enum, I think that for a lib it's a way better for the user to be able to filter on errors."
And I still think that's a good idea. 
The way you did is quite adapted for a cli tool but not for a lib.

**Comment by est31** [comment]
@calixteman oh yeah it would probably be useful to know the phase in the error. However, errors from all phases will likely include locations and error messages and IMO it's better to share them. So I've added a phase enum. If we want to pass any phase specific info, it can be put into that enum. Does that sound good?

**Comment by est31** [comment]
I can also create a totally separate error enum like you suggested, but only want to do that after you confirmed that the light enum approach the PR currently has is not enough.

**Comment by calixteman** [comment]
I think it'll be better to have an enum containing the different type of error and each of them could have some parameters.
For example, InvalidChar(Foo { c: char, loc: Loc}) and the display would be "Lexer: Invalid char  {}", c...
So when an error raised it's possible to filter it using a match on the enum. With what you propose, we must filter on the string and I think it isn't nice.

**Comment by est31** [comment]
Oh you meant that. I've adjusted the error handling to do what you proposed. Can you check again? Thanks.

File: benches/basic/ascii.cpp [hunk 1]
```diff
@@ -1,3 +1,6 @@
+#endif
```
Comment (calixteman) [inline benches/basic/ascii.cpp]: why this change ?

File: src/errors.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+use crate::lexer::lexer::Location;
```
Comment (calixteman) [inline src/errors.rs:6]: nit: add license header

File: src/errors.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+use crate::lexer::source::FileId;
+use crate::lexer::errors::LexerError;
+
+pub type Span = (Option<FileId>, Location, Location);
```
Comment (calixteman) [inline src/errors.rs]: define a struct `Span`

File: src/lexer/lexer.rs [hunk 4]
```diff
@@ -713,7 +725,13 @@ impl<'a, PC: PreprocContext> Lexer<'a, PC> {
         let id = unsafe { std::str::from_utf8_unchecked(&self.buf.slice(spos)) };
         if let Some(keyword) = PREPROC_KEYWORDS.get(id) {
             if eval {
-                self.preproc_parse(keyword.clone())
+                match self.preproc_parse(keyword.clone()) {
```
Comment (calixteman) [inline src/lexer/lexer.rs]: nit: use unwrap_or

File: src/lexer/preprocessor/preprocessor.rs [hunk 5]
```diff
@@ -166,12 +167,24 @@ impl<'a, PC: PreprocContext> Lexer<'a, PC> {
                 Token::PreprocPragma
             }
             Token::PreprocError => {
-                skip_until!(self, b'\n');
-                self.buf.add_new_line();
-                Token::PreprocError
+                let mut buf = Vec::new();
```
Comment (calixteman) [inline src/lexer/preprocessor/preprocessor.rs]: Unrelated change so please do it an other commit.

**Review by calixteman — COMMENTED** [review]
Did you run clippy ?

File: benches/basic/ascii.cpp (hunk 1 above)
Comment (est31) [reply benches/basic/ascii.cpp]: I'll remove it.

**Comment by est31** [comment]
> Did you run clippy ?

Even running rustc on the repo currently gives lots of warnings. And clippy complains about a lot of stuff too. If you want, I can submit a PR that fixes or allows the rustc and clippy warnings.

File: src/lexer/preprocessor/preprocessor.rs (hunk 5 above)
Comment (est31) [reply src/lexer/preprocessor/preprocessor.rs]: This new code is mainly there to get the error message of the `#error something` pragma. How should I put it into a separate commit? Should I first add the `#error` error without reading the message and then add the message later, or should I make one separate commit about the `#error` error?

**Comment by calixteman** [comment]
> > Did you run clippy ?
> 
> Even running rustc on the repo currently gives lots of warnings. And clippy complains about a lot of stuff too. If you want, I can submit a PR that fixes or allows the rustc and clippy warnings.

Yep you're right, it's a stupid question.

File: src/lexer/preprocessor/preprocessor.rs (hunk 5 above)
Comment (calixteman) [reply src/lexer/preprocessor/preprocessor.rs]: Ok let's do it here.
Here I'd save the pos: `let spos = self.buf.pos();`
and then skip_until '\n'.
Finally get the slice with `self.buf.slice(spos)` and then convert it into a string.

**Comment by calixteman** [comment]
Could you add some tests to test # error for example and the stringification of error ?

File: src/lexer/lexer.rs [hunk 6]
```diff
@@ -1011,6 +1031,14 @@ impl<'a, PC: PreprocContext> Lexer<'a, PC> {
         }
     }
 
+    pub(super) fn span(&self) -> Span {
+        Span {
+            file: self.buf.get_source_id(),
+            start: self.location(),
```
Comment (calixteman) [inline src/lexer/lexer.rs]: it's strange to have the same value for start & end

File: src/lexer/preprocessor/preprocessor.rs [hunk 7]
```diff
@@ -166,12 +167,14 @@ impl<'a, PC: PreprocContext> Lexer<'a, PC> {
                 Token::PreprocPragma
             }
             Token::PreprocError => {
+                let spos = self.buf.pos();
                 skip_until!(self, b'\n');
-                self.buf.add_new_line();
-                Token::PreprocError
+                let sl = self.buf.slice(spos);
+                let msg = String::from_utf8_lossy(&sl).to_string();
+                return Err(LexerError::ErrorDirective { sp: self.span(), msg });
```
Comment (calixteman) [inline src/lexer/preprocessor/preprocessor.rs:174]: use token start and current location for the span

**Review by calixteman — COMMENTED** [review]
Just add a test for the `#error` and fix the few nits and it should be good to be merged.

**Comment by est31** [comment]
Ready for another review @calixteman

File: src/lexer/lexer.rs [hunk 8]
```diff
@@ -1011,9 +1036,17 @@ impl<'a, PC: PreprocContext> Lexer<'a, PC> {
...
+            file: self.buf.get_source_id(),
+            start: self.start.unwrap(),
+            end: self.location(),
+        }
+    }
+
     pub fn next_token(&mut self) -> LocToken {
         loop {
-            let start = self.location();
+            self.start = Some(self.location());
```
Comment (calixteman) [inline src/lexer/lexer.rs]: Why are you changing this ?
And I don't see the point to have this value in an option to unwrap it after.

Comment (est31) [reply src/lexer/lexer.rs]: I'm storing start in the lexer so that I don't have to pass it as arg to whatever function that wants to return errors. It's inside an option because when constructing the lexer it's empty. I guess the unwrap will be optimized out anyways.

Comment (calixteman) [reply src/lexer/lexer.rs]: Please do it in another commit (could be merged before this one).
Explain clearly the rational and give a proof the the unwrap will be optimized.

Comment (est31) [reply src/lexer/lexer.rs]: See #9

**Comment by est31** [comment]
Rebased. Ready for review.

**Comment by calixteman** [comment]
Thx
